test/e2e/framework/framework.go

```diff
@@ -60,3 +66,64 @@ func queryAndUpdateGlobalnetStatus() {
 		return true, "", nil
 	})
 }
+
+func (f *Framework) AwaitForGatewayWithStatus(cluster framework.ClusterIndex,
+	name string, status submarinerv1.HAStatus) *submarinerv1.Gateway {
+	gwClient := SubmarinerClients[cluster].SubmarinerV1().Gateways(framework.TestContext.SubmarinerNamespace)
+	gw := framework.AwaitUntil("await gateway ready",
+		func() (interface{}, error) {
+			resGw, err := gwClient.Get(name, metav1.GetOptions{})
+			if apierrors.IsNotFound(err) {
+				return nil, nil
+			}
+			return resGw, err
+		},
+		func(result interface{}) (bool, string, error) {
+			if result == nil {
+				return false, "gateway not found yet", nil
+			}
+			gw := result.(*submarinerv1.Gateway)
+			if gw.Status.HAStatus != status {
+				return false, "", fmt.Errorf("Gateway %q showed up but has wrong status %q, expected %q",
+					gw.Name, gw.Status.HAStatus, status)
+			}
+			return true, "", nil
+		})
+	return gw.(*submarinerv1.Gateway)
+}
+
+func (f *Framework) AwaitForGatewayGone(cluster framework.ClusterIndex, name string) {
+	gwClient := SubmarinerClients[cluster].SubmarinerV1().Gateways(framework.TestContext.SubmarinerNamespace)
+	framework.AwaitUntil("await gateway gone",
+		func() (interface{}, error) {
+			if _, err := gwClient.Get(name, metav1.GetOptions{}); apierrors.IsNotFound(err) {
+				return true, nil
+			}
+			return false, nil
+		},
+		func(result interface{}) (bool, string, error) {
+			gone := result.(bool)
+			return gone, "", nil
+		})
+
+}
+
+// gatewayCleanup ensures that only the active gateway node is flagged as gateway node
+//                which could not be after a failed test which left the system on an
+//                unexpected state
+func (f *Framework) gatewayCleanup() {
+	gatewayClient := SubmarinerClients[framework.ClusterA].SubmarinerV1().Gateways(
+		framework.TestContext.SubmarinerNamespace)
+	gwList, err := gatewayClient.List(metav1.ListOptions{LabelSelector: submarinerv1.HAStatusPassiveSelector})
+	Expect(err).NotTo(HaveOccurred())
```

Comment:
By querying for `Gateway` with a certain status label for cleanup, it seems we're depending on the code we're trying to test to be correct.

Is such cleanup really necessary?  I realize one test case failure  could could result other test failures but I think that's OK - such tests have initial sanity checks which should make it clear. Also if a bug causes one test failure it may cause others anyway.

Reply:
I wrote this out of necessity, while developing the E2E tests were leaving my system on inconsistent state.

In the end many e2e tests depend on the underlaying code behaving on a certain way. It's true that if the reporting does not work, the cleanup won't work. I prefer to risk the cleanup not working ok if we break that, but having a cleanup that in most cases will prevent that the behavior of some failed tests will cascade into others.